### PR TITLE
User api

### PR DESCRIPTION
**Pull request by evanplaice**
The new User model is ready.

fname/lname have been replaced with username
'nick' is used for social media profile usernames

*Note: name can't be used because it conflicts with function.name in JS*

This can be pulled either now if somebody needs it or after I finish the User API.

### CODE REVIEW

File: test/models/user.js [hunk 1]
```diff
@@ -15,16 +15,40 @@ module.exports = (db) => {
 
     it('saves without error', (done) => {
       let user = new Model({
-        fname: 'John',
-        lname: 'Doe'
+        username: 'Marcus Aurelius',
+        email: 'marcusaurelius@rome.gov',
+        github: {
+          nick: '@soicismftw',
+          link: 'https://github.com/soicismftw'
```
Comment (SoullessWaffle) [inline test/models/user.js]: Typo? Github should probably be 'stoicismftw'.

**Comment by SoullessWaffle** [comment]
Nice work.

File: test/models/user.js (hunk 1 above)
Comment (evanplaice) [inline test/models/user.js]: Ahhh, damnit. I'll sneak in a fix on a later commit.

**Comment by thebergamo** [comment]
@evanplaice you remove the `scripts/test` in the `package.json` ? This run the tests and set the `NODE_ENV`to 'test`.

**Comment by ericdouglas** [comment]
@thebergamo `nick` to save some bits :stuck_out_tongue_winking_eye: 

For me it's ok `nick`.

**Comment by evanplaice** [comment]
@thebergamo No worries, I fixed it. Haven't pushed those changes yet tho.

**Comment by evanplaice** [comment]
I held off on adding the rest of the tests because @thebergamo mentioned using Hapi + Lab for API testing.

The currently implemented API tests use a live server/client setup. Initially, it was a pain to figure out. It required a few unique workarounds to get the express listener up, and then getting it to complete within a reasonable timeframe (ie so it wouldn't trigger Mocha's timeout limit). Seeing it in action, I'm not completely convinced that switching to Hapi + Lab would add any benefit.

A note was added to the task to add Controller tests. I'm not sure if it's necessary since the API routes and controller CRUD methods are intrinsically tied together.

Lastly, if any naming schemes need to change now is the time to come to a consensus.

**Comment by oscar-g** [comment]
Great work, @evanplaice.

The tests look great so far.  
1) I wonder if we have to start the server before each test. Why not just start it `before` the 'User API' suite? It would make the tests run faster, probably.

2)  I noticed you're using `request` to test the server routes.  We had previously thought of using [supertest](https://github.com/visionmedia/supertest) (its in the package.json).  I think it would make the tests more readable if we used supertest...take a look at the repo and let me know what you think.

3) One more thing that's an easy fix: 
The npm start script is this: `semistandard && scripts/test`
The `scripts/test` script is this: `semistandard && node_modules/.bin/mocha`
It looks like we're using running semistandard twice.

After we merge we'll need to complete the API:
* Add more test cases
* Add input sanitation
* Create docs in the github wiki.

File: controllers/user.js [hunk 2]
```diff
@@ -15,33 +15,56 @@ var Model;
...
-UserController.prototype.update = (req, res) => {
-  res.status(200).send('ok');
-};
-
-UserController.prototype.destroy = (req, res) => {
-  res.status(200).send('ok');
+  UserController.prototype.destroy = (req, res) => {
+    Model.remove(req.params.id, (err) => {
+      if (err) {
+        res.status(400).send(err);
```
Comment (007lva) [inline controllers/user.js]: This not needs a ```return``` or a ```else``` block?

**Comment by evanplaice** [comment]
@oscar-g 

Thanks. Great feedback. To answer your questions.

> 1) I wonder if we have to start the server before each test. Why not just start it before the 'User API' suite? It would make the tests run faster, probably.

Using a live server is tricky because once `server.listen()` is called it blocks until `server.close()` (ie the `done()` on `before()` doesn't execute until all the tests complete) . I initially tried inserting `server.listen()` into `before()` and `server.close()` into end. It didn't work because, from Mocha's perspective, the `before()` statement was timing out (ie taking longer than 2000ms to return). By shortening the time the server is blocking to the length that each test runs, Mocha timeouts are no longer an issue.

Now, with that said. If there's a way to move the server.listen() and server.close() calls outside of the tests altogether then timeouts becomes a non-issue. I currently can't think of a way to make that work.

> 2) I noticed you're using request to test the server routes. We had previously thought of using supertest (its in the package.json). I think it would make the tests more readable if we used supertest...take a look at the repo and let me know what you think.

I have no problem switching to Supertest. I just wanted to get an initial test up and running without adding too much complexity. I used `request` because it makes it very easy to get a http client up and running. Now that we have an initial test that works, I think it's safe to transition over to something better.

> 3) One more thing that's an easy fix: 
The npm start script is this: semistandard && scripts/test
The scripts/test script is this: semistandard && node_modules/.bin/mocha
It looks like we're using running semistandard twice.

Merged. Haha, I was wondering why the test setup was taking so long. I guess I should probably pay closer attention when I apply quick fixes.

File: controllers/user.js (hunk 2 above)
Comment (evanplaice) [inline controllers/user.js]: @007lva Good catch. The fix is incorporated in [d5a47cc](https://github.com/evanplaice/ossu-api/commit/d5a47cc1dcaf87245fda772368d29bb6a44190ae).

**Comment by oscar-g** [comment]
Good work, guys! :)

**Comment by evanplaice** [comment]
:+1:

**Comment by 007lva** [comment]
:+1:

**Comment by thebergamo** [comment]
+1 great job!

**Pull request merged**